File: app/src/main/java/com/keiferstone/dndinitiativetracker/Character.java
package com.keiferstone.dndinitiativetracker;


import android.os.Parcel;
import android.os.Parcelable;

import java.util.UUID;

class Character implements Parcelable {
    private String id;
    private String name;
    private int modifier;
    private int d20;
    private boolean marked;

    Character(String name, int modifier, int d20) {
        this.id = UUID.randomUUID().toString();
        this.name = name;
        this.modifier = modifier;
        this.d20 = d20;
        this.marked = false;
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    int getModifier() {
        return modifier;
    }

    void setModifier(int modifier) {
        this.modifier = modifier;
    }

    int getD20() {
        return d20;
    }

    void setD20(int d20) {
        this.d20 = d20;
    }

    boolean isMarked() {
        return marked;
    }

    void setMarked(boolean marked) {
        this.marked = marked;
    }

    int getInitiative() {
        return getModifier() + getD20();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Character)) return false;

        Character character = (Character) o;

        return getId().equals(character.getId());

    }

    @Override
    public int hashCode() {
        return getName().hashCode();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.id);
        dest.writeString(this.name);
        dest.writeInt(this.modifier);
        dest.writeInt(this.marked ? 1 : 0);
    }

    private Character(Parcel in) {
        this.id = in.readString();
        this.name = in.readString();
        this.modifier = in.readInt();
        this.marked = in.readInt() == 1;
    }

    public static final Parcelable.Creator<Character> CREATOR = new Parcelable.Creator<Character>() {
        @Override
        public Character createFromParcel(Parcel source) {
            return new Character(source);
        }

        @Override
        public Character[] newArray(int size) {
            return new Character[size];
        }
    };
}

File: app/src/main/java/com/keiferstone/dndinitiativetracker/Character.kt
package com.keiferstone.dndinitiativetracker

import android.os.Parcel
import android.os.Parcelable
import java.util.*

data class Character(
        var name: String,
        var modifier: Int) : Parcelable {
    var id: String = UUID.randomUUID().toString()
    var d20: Int = 1
    var isMarked: Boolean = false
    var isDead: Boolean = false
    var isSelected: Boolean = false
    val initiative: Int get() = modifier + d20

    constructor(character: Character): this(character.name, character.modifier) {
        this.id = character.id
        this.d20 = character.d20
        this.isMarked = character.isMarked
        this.isDead = character.isDead
        this.isSelected = character.isSelected
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Character) return false
        val character = other as Character?
        return id == character?.id
    }

    override fun hashCode(): Int { return id.hashCode() }

    override fun describeContents(): Int { return 0 }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.name)
        dest.writeInt(this.modifier)
        dest.writeString(this.id)
        dest.writeInt(this.d20)
        dest.writeInt(if (this.isMarked) 1 else 0)
        dest.writeInt(if (this.isDead) 1 else 0)
        dest.writeInt(if (this.isSelected) 1 else 0)
    }

    private constructor(parcelIn: Parcel): this(
            parcelIn.readString(), parcelIn.readInt()) {
        id = parcelIn.readString()
        d20 = parcelIn.readInt()
        isMarked = parcelIn.readInt() != 0
        isDead = parcelIn.readInt() != 0
        isSelected = parcelIn.readInt() != 0
    }

    companion object {
        @Suppress("unused")
        val CREATOR = createParcel { Character(it) }
    }
}
